docs/tendermint-core/p2p/peer_manager.md

```diff
@@ -0,0 +1,439 @@
+[`Dialed`](#dialed-transition) or [`Accepted`](#accepted-transition) transitions,
+providing to the peer manager a list of channels supported by the peer.
+This information is broadcast to all reactors in a `PeerUpdate` message that
+informs the new state (up) of the peer.
+
+This transition is not represented in the picture because it does not change
+the state of the peer, which should be in the [`Connected`](#connected-peer) state.
+
+### Disconnected transition
+
+This transition is performed when the node disconnects from a peer.
+It is invoked by the router when an error is returned by the routines used to
+exchange messages with the peer.
+
+The peer is expected to be in the [`Connected`](#connected-peer) state.
+If the [`Ready`](#ready-transition) transition has been performed, the peer manager broadcast a
+`PeerUpdate` to all reactors notifying the new status (down) of this peer.
+
+If the peer is still present in the peer store, its `LastDisconnected` time is
+set and the peer manager spawns a routine that after `DisconnectCooldownPeriod`
+notifies the next peer to dial routine about the availability of this peer.
+Until then, the peer is the `Frozen Candidate` state.
+
+### Errored transition
+
+This transition is performed when a reactor interacting with the peer reports
+an error to the router.
+
+The peer is expected to be in the [`Connected`](#connected-peer) state.
+If so, the peer transitions to the [`Evict`](#evict-peer) sub-state, which
+should lead the router to disconnect from the peer, and the next peer to evict
+routine is notified.
+
+### Upgrading peer
+
+A [connected peer](#connected-peer) which should be evicted to given room to a
+higher-ranked peer the router is dialing to.
+
+The `Upgrading` sub-state is part of the procedure to [upgrade connection slots](#slots-upgrade).
+When a connection with the higher-ranker peer that should take the connection
+slot from this peer is [established](#dialed-transition), the
+[eviction](#evict-peer) of this peer is scheduled.
+
+### Evict peer
+
+A peer whose eviction was scheduled, for either of the following reasons:
+
+1. to given room to a higher-ranked peer the router is connected to, as part of
+   the procedure to [upgrade connection slots](#slots-upgrade),
+2. or because this peer was reported as [errored](#errored-transition) by a
+   reactor interacting with this peer.
+
+This peer is a [connected peer](#connected-peer).
+`Evict` is the first sub-state of the procedure that should lead the node to
+[disconnect](#disconnected-transition) from a peer.
+
+### EvictNext transition
+
+This transition returns a peer to the router to evict.
+
+The state transition is performed whenever the peer manager has scheduled the
+eviction of a peer, i.e., whenever there is a peer on `Evict` sub-state.
+The peer to evict must be a peer in the `Connected` state.
+
+The peer to evict is randomly picked from the possible multiple peers with
+eviction scheduled.
```

Comment:
I realize that peers are evicted one by one (serialized evictions). Is that the case for dialling as well? Maybe a paragraph towards the beginning of this file would help clarify this.

Reply:
Yes, it is also the case of dialing. However, the procedures to dial and evict peers are typically not blocking. In the case of dialing, there is a poll of dialing threads, meaning that we only block if all threads in this pool are busy (*). In the case of evicting, the router essentially just closes a channel, which should be fast.

(*) Notice that we increased a lot the number of threads in this pool in #8827. Moreover, there were a sleep between dialing new peers in the original designed, which was removed in #8839. As a result, we can say that, while peers to dial are provided one by one, we actually dial to several peers in parallel.

Reply:
Thanks for the explanation. I get it now. Consider adding this somewhere in the text

Reply:
Added to the router spec, including the two PRs.